### PR TITLE
Implement GRAPHITE attack, fixed

### PR DESCRIPTION
**Pull request by ryan-feng**
# Description

Add GRAPHITE attack (https://arxiv.org/pdf/2002.07088.pdf, https://github.com/ryan-feng/GRAPHITE)

Fixes #1751 

## Type of change

Please check all relevant options.

- [ ] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [X] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- [X] Tested with notebooks
- [X] Independently tested for correctness
- [ ] Test cases added, did not know how to run outside of the PR

**Test Configuration**:
- OS - Ubuntu
- Python 3.8
- ART version or commit number: dev_1.12.0
- Torch version 1.7

# Checklist

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code
- [X] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: art/attacks/evasion/graphite/graphite_blackbox.py [hunk 1]
```diff
@@ -0,0 +1,896 @@
...
+    ) -> None:
+        """
+        Create a GRAPHITEBlackbox attack instance.
+
+        :param classifier: A trained classifier.
+        :param noise_size: The resolution to generate perturbations in (w, h).
+        :param net_size: The resolution to resize images to before feeding to the model in (w, h).
+        :param heat_patch_size: The size of the heatmap patches in (w, h).
+        :param heat_patch_stride: The stride of the heatmap patching in (w, h).
+        :param heatmap_mode: The mode of heatmap in ['Target', 'Random'].
```
Comment (keykholt) [inline art/attacks/evasion/graphite/graphite_blackbox.py:143]: This is a bit vague. I assume this refers to the heatmap generation method. The values can be ['Target', 'Random']. A minor suggestion would be to not make this variable case sensitive.

File: art/attacks/evasion/graphite/graphite_blackbox.py [hunk 2]
```diff
@@ -0,0 +1,896 @@
...
+                    i : min(i + self.heat_patch_size[1], mask.shape[0]),
+                    j : min(j + self.heat_patch_size[0], mask.shape[1]),
+                ] = patch
+                new_mask = new_mask * mask
+                if np.sum(new_mask) > 0:
+                    patches.append(new_mask)
+                    indices.append((i, j))
+
+        # compute heatmap and order
+        if self.heatmap_mode == "Random":
```
Comment (keykholt) [inline art/attacks/evasion/graphite/graphite_blackbox.py:433]: One place to make case-insensitive if you decide to make a change to the heatmap_mode

File: art/attacks/evasion/graphite/graphite_whitebox_pytorch.py [hunk 3]
```diff
@@ -0,0 +1,578 @@
...
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+"""
+This module implements the white-box version of the GRAPHITE attack `GRAPHITEWhiteboxPyTorch`.
```
Comment (keykholt) [inline art/attacks/evasion/graphite/graphite_whitebox_pytorch.py:41]:
```suggestion
This module implements the white-box PyTorch version of the GRAPHITE attack `GRAPHITEWhiteboxPyTorch`.
```

File: art/attacks/evasion/graphite/graphite_whitebox_pytorch.py [hunk 4]
```diff
@@ -0,0 +1,578 @@
...
+if TYPE_CHECKING:
+    from art.utils import CLASSIFIER_NEURALNETWORK_TYPE
+    import torch
+
+logger = logging.getLogger(__name__)
+
+
+class GRAPHITEWhiteboxPyTorch(EvasionAttack):
+    """
+    Implementation of the white-box GRAPHITE attack from Feng et al. (2022). This is a physical attack
```
Comment (keykholt) [inline art/attacks/evasion/graphite/graphite_whitebox_pytorch.py:70]:
```suggestion
    Implementation of the white-box PyTorch GRAPHITE attack from Feng et al. (2022). This is a physical attack
```

**Review by keykholt — CHANGES_REQUESTED** [review]

**Comment by keykholt** [comment]
Looks good, I think I just want to clarify that this only seems to implement a whitebox attack for pytorch and not other estimator types.

File: art/attacks/evasion/graphite/graphite_whitebox_pytorch.py [hunk 5]
```diff
@@ -0,0 +1,578 @@
...
+        "off_x_range",
+        "off_y_range",
+        "blur_kernels",
+        "batch_size",
+    ]
+    _estimator_requirements = (BaseEstimator, ClassifierMixin)
+
+    def __init__(
+        self,
+        classifier: "CLASSIFIER_NEURALNETWORK_TYPE",
```
Comment (keykholt) [inline art/attacks/evasion/graphite/graphite_whitebox_pytorch.py:101]: This seems to only support PyTorchClassifiers based on the inputs to the other methods?

**Review by keykholt — CHANGES_REQUESTED** [review]

File: art/attacks/evasion/graphite/graphite_whitebox_pytorch.py (hunk 5 above)
Comment (ryan-feng) [inline art/attacks/evasion/graphite/graphite_whitebox_pytorch.py:101]:
```suggestion
        classifier: "PyTorchClassifier",
```

File: art/attacks/evasion/graphite/graphite_whitebox_pytorch.py [hunk 6]
```diff
@@ -0,0 +1,578 @@
...
+
+from art.config import ART_NUMPY_DTYPE
+from art.attacks.attack import EvasionAttack
+from art.estimators.estimator import BaseEstimator
+from art.estimators.classification import ClassifierMixin
+from art.utils import compute_success, to_categorical, check_and_transform_label_format, is_probability
+from art.attacks.evasion.graphite.utils import convert_to_network, get_transform_params, transform_wb
+
+if TYPE_CHECKING:
+    from art.utils import CLASSIFIER_NEURALNETWORK_TYPE
```
Comment (ryan-feng) [inline art/attacks/evasion/graphite/graphite_whitebox_pytorch.py:62]:
```suggestion
    from art.estimators.classification.pytorch import PyTorchClassifier
```

**Review by keykholt — APPROVED** [review]

**Review by beat-buesser — APPROVED** [review]
Hi @ryan-feng Thank you very much for contributing GRAPHITE to ART!

@keykholt Thank you very much for supporting and reviewing this PR!

**Pull request merged**